src/main/scala/uk/gov/hmrc/stub/dynamic/repository/DynamicTestDataMongoRepository.scala

```diff
@@ -101,33 +102,66 @@ class DynamicTestDataRepository @Inject() ()(implicit mongo: ReactiveMongoCompon
   protected def findByTestId(testId: String): JsObject = Json.obj("expectation.testId" -> testId)
 
   private def modifierForInsert(uriSeq: Seq[URI], expectation: ExpectationMongo): JsObject = {
-    val delay = optionalValues(expectation.delay, "expectation.delay")
-    val resultCode = optionalValues(expectation.resultCode, "expectation.resultCode")
-    Json.obj(
-      "$setOnInsert" -> Json.obj("expiry" -> (DateTimeUtils.now.getMillis + expectation.toLive.toMillis)),
-      "$setOnInsert" -> Json.obj("expectation.testId" -> expectation.testId),
-      "$set" -> Json.obj("expectation.template" -> expectation.template),
-      "$set" -> Json.obj("uri" -> Json.arr(uriSeq.map(uri => uri.toASCIIString)))
-    ) ++ delay ++ resultCode
+
+    val fieldUpdateOperators = List(
+      FieldOperator(SetOnInsertOp, Json.obj("expiry" -> (DateTimeUtils.now.getMillis + expectation.toLive.toMillis))),
+      FieldOperator(SetOnInsertOp, Json.obj("expectation.testId" -> expectation.testId)),
+      FieldOperator(SetOp, Json.obj("expectation.template" -> expectation.template)),
+      FieldOperator(SetOp, Json.obj("uri" -> uriSeq.map(uri => uri.toASCIIString))),
+      setOrUnsetOperator(expectation.delay, "expectation.delay"),
+      setOrUnsetOperator(expectation.resultCode, "expectation.resultCode")
+    )
+
+    // collect together all like operators to make a modifier Json document
+    val groups: Map[String, JsObject] =
+      fieldUpdateOperators
+        .groupBy(_.fieldOp)
+        .map { op: (FieldOp, List[FieldOperator]) =>
+          op._1.toString -> op._2.foldLeft(Json.obj())(_ ++ _.jsObject)
+        }
```

Comment:
note: could do something like 
```suggestion
        .map { case (operator, document) =>
          operator.toString -> document.foldLeft(Json.obj())(_ ++ _.jsObject)
        }
```
to avoid the ugly `_1` and `_2` calls

Reply:
Yeah, it looked even worse before I put those case classes in.  I agree it's a bit horrible.  I needed to find a way to create a dynamic set of operations and then group by operator.  I don't like _1 or _2 either.  I'll try your suggestion, thanks.